module2/gift-certificate-system/web/src/main/java/com/epam/esm/controller/TagController.java

```diff
@@ -9,38 +12,95 @@
         this.tagService = tagService;
+        this.certificateService = certificateService;
         this.tagValidator = tagValidator;
     }
 
-    @PatchMapping(path = "/{id}", consumes = "application/json-patch+json")
-    public ResponseEntity<Integer> updateGiftCertificate(@PathVariable Long id, @RequestBody JsonPatch patch) {
+    @GetMapping(value = "/tags")
+    public List<Tag> findAllTags() {
+        return tagService.findAllTags();
+    }
+
+    @GetMapping(value = "tags/{id}")
+    public Tag findTagById(@PathVariable Long id) {
+        return tagService.findTagById(id);
+    }
+
+    @PostMapping(path = "/tags", consumes = "application/json", produces = "application/json")
+    public ResponseEntity<Tag> addTag(@RequestBody @Valid Tag tag,
+                                                              UriComponentsBuilder ucBuilder) {
+
+        BindingResult result = new BeanPropertyBindingResult(tag, "tag");
+        tagValidator.validate(tag, result);
+        if (result.hasErrors()) {
+            return new ResponseEntity(result.getAllErrors(), HttpStatus.BAD_REQUEST);
+        } else {
+            Long tagId = tagService.saveTag(tag);
+            HttpHeaders headers = new HttpHeaders();
+            headers.setLocation(ucBuilder.path("/tags/{id}").buildAndExpand(tagId).toUri());
+            return new ResponseEntity<>(headers, HttpStatus.CREATED);
+        }
+    }
+
+    @PatchMapping(path = "tags/{id}", consumes = "application/json-patch+json")
+    public ResponseEntity<Tag> updateTag(@PathVariable Long id,
+                                                         @RequestBody JsonPatch patch,
+                                                         UriComponentsBuilder ucBuilder) {
         try {
             Tag oldTag = tagService.findTagById(id);
             Tag tagPatched = applyPatchToTag(patch, oldTag);
-            tagService.updateTag(tagPatched);
-            return ResponseEntity.ok().build();
+            BindingResult result = new BeanPropertyBindingResult(tagPatched, "tagPatched");
+            tagValidator.validate(tagPatched, result);
+            if (result.hasErrors()) {
+                return new ResponseEntity(result.getAllErrors(), HttpStatus.BAD_REQUEST);
+            } else {
+                tagService.updateTag(tagPatched);
+                HttpHeaders headers = new HttpHeaders();
+                headers.setLocation(ucBuilder.path("/tags/{id}").buildAndExpand(tagPatched.getId()).toUri());
+                return new ResponseEntity<>(headers, HttpStatus.CREATED);
+            }
         } catch (JsonPatchException | JsonProcessingException e) {
             return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
         } catch (Exception e) {
             return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
         }
     }
 
+    @DeleteMapping(path = "tags/{id}")
+    public ResponseEntity<Void> deleteGiftCertificate(@PathVariable Long id) {
```

Comment:
why you delete gift certificate inside TagController?
may be it will be better to Delete Sertificate inside CertificateController?